dev/tests/api-functional/testsuite/Magento/GraphQl/Quote/Customer/SetShippingMethodsOnCartTest.php

```diff
@@ -45,108 +66,461 @@ class SetShippingMethodsOnCartTest extends GraphQlAbstract
     protected function setUp()
     {
         $objectManager = Bootstrap::getObjectManager();
+        $this->productRepository = $objectManager->get(ProductRepositoryInterface::class);
+        $this->rateFactory = $objectManager->get(RateFactory::class);
         $this->quoteResource = $objectManager->get(QuoteResource::class);
         $this->quoteFactory = $objectManager->get(QuoteFactory::class);
         $this->quoteIdToMaskedId = $objectManager->get(QuoteIdToMaskedQuoteIdInterface::class);
+        $this->quoteRepository = $objectManager->get(CartRepositoryInterface::class);
         $this->customerTokenService = $objectManager->get(CustomerTokenServiceInterface::class);
     }
 
+    /**
+     * @magentoApiDataFixture Magento/Checkout/_files/quote_with_virtual_product_and_address.php
+     * @throws \Exception
+     */
     public function testShippingMethodWithVirtualProduct()
     {
-        $this->markTestIncomplete('https://github.com/magento/graphql-ce/issues/423');
+        $maskedQuoteId = $this->getMaskedQuoteIdByReversedQuoteId('test_order_with_virtual_product');
+
+        /** @var Quote $quote */
+        $quote = $this->quoteFactory->create();
+        $this->quoteResource->load($quote, 'test_order_with_virtual_product', 'reserved_order_id');
+
+        $shippingAddress = $quote->getShippingAddress();
+
+        /** @var Rate $rate */
+        $rate = $this->rateFactory->create();
```

Comment:
We should not operate with Rate object
Tests should contain just what we want to test

1) Preconditions: some quote with products and saved adresses
2) Set shipping address in test
3) Check result

Examples:
`\Magento\GraphQl\Quote\Customer\SetShippingAddressOnCartTest`
`\Magento\GraphQl\Quote\Customer\SetPaymentMethodOnCartTest`
`\Magento\GraphQl\Quote\Customer\SetBillingAddressOnCartTest`